include/boost/histogram/python/histogram.hpp

```diff
@@ -6,59 +6,61 @@
-        bool underflow     = bh::axis::traits::options(h.axis(i)) & bh::axis::option::underflow;
-        ssize_t extent_dim = bh::axis::traits::extent(h.axis(i));
-        ssize_t size_dim   = h.axis(i).size();
+template <class Axes, class T>
+py::buffer_info make_buffer_impl(const Axes &axes, bool flow, T *ptr) {
+    // strides are in bytes
+    auto shape     = bh::detail::make_stack_buffer<ssize_t>(axes);
+    auto strides   = bh::detail::make_stack_buffer<ssize_t>(axes);
+    ssize_t stride = sizeof(T);
+    unsigned rank  = 0;
+    char *start    = reinterpret_cast<char *>(ptr);
+    bh::detail::for_each_axis(axes, [&](const auto &axis) {
+        const bool underflow = bh::axis::traits::options(axis) & bh::axis::option::underflow;
         if(!flow && underflow)
-            start += factor;
-        diminsions.push_back(flow ? extent_dim : size_dim);
-        strides.push_back(factor * size_of);
-        factor *= extent_dim;
-    }
+            start += stride;
+        const auto extent = bh::axis::traits::extent(axis);
+        shape[rank]       = flow ? extent : axis.size();
+        strides[rank]     = stride;
+        stride *= extent;
+        ++rank;
+    });
 
-    return py::buffer_info(&(*h.begin()) + start,                               // Pointer to buffer
-                           sizeof(in_storage_t),                                // Size of one scalar
-                           py::format_descriptor<in_storage_value_t>::format(), // Python struct-style format descriptor
-                           rank,                                                // Number of dimensions
-                           diminsions,                                          // Buffer dimensions
-                           strides                                              // Strides (in bytes) for each index
+    return py::buffer_info(start,                              // Pointer to buffer
+                           sizeof(T),                          // Size of one scalar
+                           py::format_descriptor<T>::format(), // Python format descriptor
+                           rank,                               // Number of dimensions
+                           shape,                              // Buffer shape
+                           strides                             // Strides (in bytes) for each index
     );
 }
+} // namespace detail
+
+/// Build and return a buffer over the current data.
+/// Flow controls whether under/over flow bins are present
+template <class A, class T>
+py::buffer_info make_buffer(bh::histogram<A, bh::dense_storage<T>> &h, bool flow) {
+    const auto &axes = bh::unsafe_access::axes(h);
+    auto &storage    = bh::unsafe_access::storage(h);
+    return detail::make_buffer_impl(axes, flow, &storage[0]);
+}
 
-/// Unlimited storage does not support buffer access
-template <typename A>
+/// Unlimited storage does not support buffer access yet
+template <class A>
```

Comment:
Why change to class? https://blogs.msdn.microsoft.com/slippman/2004/08/11/why-c-supports-both-class-and-typename-for-type-parameters/

Reply:
class is shorter.

Reply:
class is also visibly more distinct from `template`

Reply:
mp11 uses class, and I use class in boost::histogram.

Reply:
Type name actually describes what it is though, more closely than class. Class is there for historical reasons. Not a strong preference, I've just mostly seen and use typename for this.

Reply:
Pybind11 uses typename.

Reply:
I have a strong preference for class and I am already making a compromise by not requiring this project to follow the style of Boost::Histogram. I prefered `typename` when I was less experienced, but when you do a lot of TMP, you realize the benefits of `class` because it is shorter. pybind11 is a good library, but it cannot compete in with mp11 as a guideline, which was written by someone who is way beyond in skill of either you or me.

Reply:
I would like to merge this and not discuss style. I have another patch which depends on this patch in the pipeline.

Reply:
Okay. But in the future please follow the current style and then make a PR or issue to change the style if you don't like it. A mix of the two is much worse than just one or the other. I'll change this to use class everywhere; as I said, I don't have a strong preference. (And we may end up on the Boost::Histogram style)

The standard library (libstdc++ and libc++) uses both interchangeably, of course...